polaris-core/src/main/java/org/apache/polaris/core/persistence/BasePersistence.java

```diff
@@ -392,4 +392,15 @@ boolean hasChildren(
       @Nullable PolarisEntityType optionalEntityType,
       long catalogId,
       long parentId);
+
+  /**
+   * Creates a fresh copy of this persistence implementation that is able to function independently
+   * of {@code this object}. The {@code copyOf} method is intended to be used to isolate state of
+   * {@link BasePersistence} instances in different thread.
+   *
+   * <p>If a particular implementation is thread-safe, it may return {@code this} from this method.
+   */
+  default BasePersistence copyOf() {
+    return this;
```

Comment:
The name `copyOf` is confusing when we allow it to return itself. I'd recommend to either give a name with wider scope or we don't provide a default impl.

Reply:
WDYT about `disengage()`?

Reply:
For that matter, why not implement `Cloneable`?

Reply:
Then maybe remove the default impl and let implementations return the copy - as it's done for all the other objects.

Reply:
`Cloneable` is too deeply rooted in the JRE... I prefer simple copy via explicit constructor. Will update presently.

Reply:
updated

Reply:
If the base idea behind this PR is to actually provide a deep clone of `PolarisCallContext`, I'd vote to rename all `copyOf` methods to something like `deepClone` or `deepCopy`. But since that's largely a matter of taste, I'm fine with `copyOf` as well.

Reply:
This specific change is needed for the upcoming NoSQL persistence.

I'm open to call context refactoring too, but I guess we may want to handle thread locals first?.. Then, perhaps deep clone will not be necessary :thinking: Cf. #463